### PR TITLE
[Refactor:InstructorUI] Change Markdown Engine to Commonmark

### LINKED ISSUE #4208: Better Code Boxes for Forum
Looking at https://github.com/Submitty/Submitty/pull/4186 which was merged recently, I'd like to make a request to not wrap lines of code, but rather add a horizontal scroll bar. This is how I've seen most implementations for code viewers, and it's especially useful given the lack of line numbers. I think this is especially necessary for languages where whitespace is significant.

It may be that there was already a design discussion offline about this, and if so, I'd like to see a high level summary (as the original issue calls for scrolling) and then feel free to close this issue.

The wishlist:

* grey background for the block of code
* language specific syntax highlighting (perhaps a default per course, but allow author to specify/override)
* line numbers
* [probably] use horizontal scrolling for extra wide lines (don't wrap)

### PR DESCRIPTION
**Pull request by eddiekrystowski**
### Please check if the PR fulfills these requirements:
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Closes #4208 
Submitty uses [Parsedown](https://github.com/erusev/parsedown) as a markdown engine. It is no longer actively maintained or updated. It also lacks customization options.

### What is the new behavior?
The markdown engine has been changed to [League\CommonMark](https://commonmark.thephpleague.com/) which follows closely with the [CommonMark](https://commonmark.org/) markdown spec. This engine also allows the definition of custom renderers/extensions, making it highly more customizable.

Due to the customization potential, markdown code blocks on Submitty (not CodeMirror) are now styled with syntax highlighting, background color, automatic language detection, line numbers, and horizontal scrolling on large lines.

image
image

### CODE REVIEW

**Review by cjreed121 — APPROVED** [review]
Works great. Not sure if you are already planning to do this but make sure the docs get updated: https://submitty.org/student/communication/markdown. There seems to be some new/different syntax for Markdown with the new engine.

File: site/composer.json [hunk 1]
```diff
@@ -29,7 +29,9 @@
     "cboden/ratchet": "0.4.3",
     "textalk/websocket": "1.5.3",
     "egulias/email-validator": "3.1.1",
-    "maennchen/zipstream-php": "^2.1"
+    "maennchen/zipstream-php": "^2.1",
+    "league/commonmark": "^1.6",
+    "spatie/commonmark-highlighter": "^2.1"
```
Comment (MasterOdin) [inline site/composer.json:34]: Use specific versions here. As Submitty isn't a library, using ranges doesn't get us anything over specific values which dependabot handles bumping.

To resolve the merge conflict though, I would recommend doing:
```bash
git checkout master
git pull master
git checkout commonmark
git add .
git commit # commit merge
wget https://raw.githubusercontent.com/Submitty/Submitty/master/site/composer.lock -O composer.lock
wget https://raw.githubusercontent.com/Submitty/Submitty/master/site/composer.json -O composer.json
cd site
composer require "league/commonmark:1.6.5"
composer require "spatie/commonmark-highlighter:2.1.1"
```
(typing this free-hand, probably some error in wget and composer usage as I can never fully remember the former's flags and the latter specifying specific version using `:` or `@`.

Comment (MasterOdin) [reply site/composer.json:34]: Additionally, can we remove the `erusev/parsedown` dependency?

**Review by MasterOdin — CHANGES_REQUESTED** [review]

File: site/composer.json (hunk 1 above)
Comment (eddiekrystowski) [reply site/composer.json]: Thanks for the help on the merge conflict. I just removed the parsedown dependency.

**Review by MasterOdin — APPROVED** [review]

**Pull request merged**